simulator/infra/macro.h

```diff
@@ -280,4 +280,74 @@ static inline uint128 gen_reverse( uint128 /* src1 */, size_t /* shamt */) {
     throw std::runtime_error( "Generalized reverse is not implemented for RV128");
     return 0;
 }
+
+uint64_t shuffle_stage( uint64_t src, uint64_t maskL, uint64_t maskR, size_t N)
+{
+    uint64_t x = src & ~( maskL | maskR);
+    x |= ( ( src << N) & maskL) | ( ( src >> N) & maskR);
+    return x;
+}
+
+// array with constants for shuffle
+std::array<uint64_t> shfl_consts[] = {
```

Comment:
Do I understand correctly it is a 2D array actually?

Reply:
Yes, the first column is the second argument of `shuffle_stage` function and the second is the third argument of this function. Now I think it would be a bit better just to pass an index to `shuffle_stage` and get constants inside it.